Track virtual documents.

NTaylorMullen opened the pull request:
- Track opening, closing and changing of virtual documents that don't exist on disk.
- Updated the various document selectors to properly include virtual files.

#2431

/cc @DustinCampbell @rchande

NTaylorMullen commented on src/omnisharp/extension.ts:
```diff
@@ -81,6 +81,7 @@ export async function activate(context: vscode.ExtensionContext, packageJSON: an
         localDisposables.add(vscode.languages.registerCodeActionsProvider(documentSelector, codeActionProvider));
         localDisposables.add(reportDiagnostics(server, advisor));
         localDisposables.add(forwardChanges(server));
```
Based a lot of these changes off of the forwardChanges handler and I didn't see any tests. What's the test guidance to test my virtual document tracker in this repo?

rchande replied:
We don't yet have a strong unit testing story for this, but it is possible to write integration tests -- take a look at the `integrationTests` folder. Not sure if there's an easy way to create virtual documents programmatically without installing another extension though. The "code action rename" test covers this type of scenario.

rchande commented on src/features/virtualDocumentTracker.ts:
```diff
@@ -0,0 +1,91 @@
+    let req = { FileName: document.uri.path, changeType: FileChangeType.Create };
+    serverUtils.filesChanged(server, [req])
+        .catch(err => {
+            console.warn(`[o] failed to forward virtual document change event for ${document.uri.path}`, err);
```
@akshita31 Can you weigh in on the best way to do this logging?

akshita31 replied:
Yes, we should use the C# channel to do the logging here.
@NTaylorMullen The way logging works in the extension , is that we post a  [loggingEvent](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/omnisharp/loggingEvents.ts) to an [https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/EventStream.ts](event stream), to which many observers have susbsribed. When we post the event to the event stream, all the observers that have subscribed to the stream, perform the necessary action. For this case, we will have to add the appropriate logging event and make the [C# logger observer](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/observers/CsharpLoggerObserver.ts) listen to it. I am not sure if the [C# channel observer ](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/observers/CsharpChannelObserver.ts) , should care of these events and show the channel. I will leave that up to you.

rchande replied:
@akshita31 How does this component get access to the eventStream?

akshita31 replied:
We can pass it down from the extension.ts - https://github.com/OmniSharp/omnisharp-vscode/pull/2436/files#diff-f49a39a20dc9807208c37823c89847a6R84

rchande commented on src/features/virtualDocumentTracker.ts:
```diff
@@ -0,0 +1,91 @@
+
+function openVirtualDocument(document: TextDocument, server: OmniSharpServer) {
+    let req = { FileName: document.uri.path, changeType: FileChangeType.Create };
+    serverUtils.filesChanged(server, [req])
```
filesChanged and updateBuffer are both async/promise returning and should be awaited

NTaylorMullen replied:
I thought so as well but that's not done in the `changeForwarding` class. Was trying to be consistent. Is it a mistake that it's not awaited there?

rchande replied:
In theory, maybe it doesn't matter because your requests get put into the queue in the submission order.

akshita31 replied:
@NTaylorMullen Recently we are trying to move to using async/await rather than .then/catch for promises. The `changeForwarding` still has the old pattern, but for this class we should prefer try/catch and async/await

rchande commented on src/features/virtualDocumentTracker.ts:
```diff
@@ -0,0 +1,91 @@
+import CompositeDisposable from '../CompositeDisposable';
+
+function trackCurrentVirtualDocuments(server: OmniSharpServer) {
+    let registration = server.onProjectAdded(() => {
```
This is going to add the existing virtual documents every time O# discovers a project, which can happen multiple times. There is no "done loading" event today, unfortunately.

rchande replied:
(We should test this in a multi-project + razor scenario and verify o# doesn't do anything weird if you add a buffer twice)

NTaylorMullen replied:
This code is only ever called once for the first project. See at the end of the method where it disposes its own registration to the method.

rchande replied:
Ah, right. Thanks.

NTaylorMullen commented:
🆙 📅 covered all feedback.

Tried writing more tests to no avail sadly. Wanted to also verify the closing text document path but that's not supported by VSCode extension APIs (the IDE does it when it wants to). Also wanted to verify the "already opened" scenario but couldn't figure out how to do that because the Omnisharp server is booted once and run throughout all of the integration tests (need to have a virtual document created before server).

NTaylorMullen commented:
Also, travis failure seems unrelated.

NTaylorMullen commented:
ping

rchande commented:
@NTaylorMullen Travis failure sounds like it could be related, given that it's in the file change notification layer, which is what's being changed here. Kind of sounds like we're trying to send a file change notification before O# has started of after it has stopped.

```
o] failed to forward file change event for /home/travis/build/OmniSharp/omnisharp-vscode/test/integrationTests/testAssets/singleCsproj/obj/project.assets.json Error: This socket is closed
    at Socket._writeGeneric (net.js:692:19)
    at Socket._write (net.js:743:8)
    at doWrite (_stream_writable.js:329:12)
    at writeOrBuffer (_stream_writable.js:315:5)
    at Socket.Writable.write (_stream_writable.js:241:11)
    at Socket.write (net.js:670:40)
    at OmniSharpServer._makeRequest (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/server.ts:37:2)
    at RequestQueue.OmniSharpServer._requestQueue.requestQueue_1.RequestQueueCollection.request [as _makeRequest] (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/server.js:8:270)
    at RequestQueue.processPending (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/requestQueue.ts:24:38)
    at RequestQueueCollection.drain (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/requestQueue.ts:24:38)
    at RequestQueueCollection.enqueue (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/requestQueue.ts:24:38)
    at Promise (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/server.ts:23:61)
    at Promise (<anonymous>)
    at OmniSharpServer.<anonymous> (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/server.ts:23:61)
    at Generator.next (<anonymous>)
    at /home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/server.js:4:44590
    at Promise (<anonymous>)
    at __awaiter (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/server.js:4:43765)
    at OmniSharpServer.makeRequest (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/src/omnisharp/server.ts:23:61)
    at Object.<anonymous> (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/utils.js:4:20562)
    at Generator.next (<anonymous>)
    at /home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/utils.js:4:18954
    at Promise (<anonymous>)
    at __awaiter (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/utils.js:4:18089)
    at Object.filesChanged (/home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/omnisharp/utils.js:4:20461)
    at /home/travis/build/OmniSharp/omnisharp-vscode/vsix/extension/out/src/features/changeForwarding.js:4:6000
    at e.fire (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:96:764)
    at /home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:479:223
    at e.fire (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:96:764)
    at e.$onFileEvent (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:480:375)
    at e._doInvokeHandler (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:636:832)
    at e._invokeHandler (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:636:550)
    at e._receiveRequest (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:635:631)
    at e._receiveOneMessage (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:635:400)
    at /home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:634:315
    at /home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:637:395
    at /home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:95:432
    at e.fire (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:96:764)
    at Socket.<anonymous> (/home/travis/build/OmniSharp/omnisharp-vscode/.vscode-test/stable/VSCode-linux-x64/resources/app/out/vs/workbench/node/extensionHostProcess.js:154:338)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
    at readableAddChunk (_stream_readable.js:178:18)
    at Socket.Readable.push (_stream_readable.js:136:10)
    at Pipe.onread (net.js:560:20)
```

rchande commented on src/features/virtualDocumentTracker.ts:
```diff
@@ -0,0 +1,97 @@
+            }
+        }
+
+        registration.dispose();
```
I'm not sure what the exact semantics of the NodeJS EventEmitter are. Is it possible for this handler to reenter? If so you have a race on disposing the registration/could process the virtual documents more than once, no?

NTaylorMullen replied:
maybe the vscode semantics are different but in normal node / JavaScript it's single threaded thus making races impossible.

NTaylorMullen replied:
Ah, actually, the await yields execution. I can see how things can get invoked more than once unintentionally.

rchande commented:
@NTaylorMullen This looks pretty good to me. If the Travis failure isn't obvious to you, we can requeue the build and see if it goes away. If you're interested in getting an "existing virtual document" tested, there are a couple of approaches that might work: 
* Add a hook to all of our existing tests that adds a document provider before we activate the C# extension
* Issue the "restart OmniSharp" command in your test after adding the document provider

NTaylorMullen commented:
Thanks for the feedback, I think you were spot on in all accounts. I've addressed all the feedback, verified tests run locally (hopefully travis passes!).

> Add a hook to all of our existing tests that adds a document provider before we activate the C# extension
> Issue the "restart OmniSharp" command in your test after adding the document provider

Definitely concerned about doing something like that, sounds too extreme. It would muck with any sort of expectations around what documents are currently open in the editor for every other test and could lead to flakeyness in the requirement to restart omnisharp. If you're alright with it I'd rather pass.

🆙 📅

rchande commented:
@NTaylorMullen Looks like we got the same build failure for your update...

NTaylorMullen commented:
> @NTaylorMullen Looks like we got the same build failure for your update...

Ya I saw that too. I pushed another master branch clone and am seeing if travis fails on that. Those failures don't occur locally.

NTaylorMullen commented:
@rchande ya same failure on the master branch (without my changes): https://travis-ci.org/OmniSharp/omnisharp-vscode/builds/410456563?utm_source=github_status&utm_medium=notification.

Will that block the inclusion of this PR?

akshita31 commented:
@NTaylorMullen Yes, currently travis is a mandatory test for merging the pull request. I am investigating further into why the builds are failing.

rchande commented:
@NTaylorMullen Thanks for investigating, we'll take a look.

akshita31 commented:
@NTaylorMullen The CI issue has been fixed. Thank you for taking the time to contribute to the repo.
